array.c

```diff
@@ -5392,6 +5392,20 @@ rb_ary_sample(rb_execution_context_t *ec, VALUE ary, VALUE randgen, VALUE nv, VA
     }
     n = NUM2LONG(nv);
     if (n < 0) rb_raise(rb_eArgError, "negative sample number");
+    if (replace) {
+        if (replace != Qtrue) {
+            rb_raise(rb_eArgError, "unexpected value for replace: %"PRIsVALUE,
+                     rb_obj_class(replace));
+        }
+        result = rb_ary_new_capa(n);
+        while (n-- > 0) {
+            i = RAND_UPTO(len);
+            len = RARRAY_LEN(ary);
```

Comment:
Why is this line necessary?

Reply:
Maybe I got the reason. Can `RAND_UPTO` can involve thread switching due to calling `rb_funcall`?

Reply:
Yes, exactly.